File: src/main/java/com/nurkiewicz/tsclass/parser/ast/Method.java
package com.nurkiewicz.tsclass.parser.ast;

import com.google.common.collect.ImmutableList;
import lombok.Value;

import java.util.stream.Collectors;

@Value
public class Method {

    private final String name;
    private final Type type;
    private final ImmutableList<Parameter> parameters;
    private final ImmutableList<Statement> statements;

    public String methodDescriptor() {
        return org.objectweb.asm.Type.getMethodDescriptor(getType().toJavaType(), getParamJavaTypes());
    }

    private org.objectweb.asm.Type[] getParamJavaTypes() {
        return getParameters()
                .stream()
                .map(Parameter::getType)
                .map(Type::toJavaType)
                .collect(Collectors.toList())
                .toArray(new org.objectweb.asm.Type[0]);
    }
}

File: src/main/kotlin/com/nurkiewicz/tsclass/parser/ast/Method.kt
package com.nurkiewicz.tsclass.parser.ast

data class Method(
        val name: String,
        val type: Type,
        val parameters: List<Parameter>,
        val body: Block) {

    fun methodDescriptor(): String {
        return org.objectweb.asm.Type.getMethodDescriptor(type.toJavaType(), *paramJavaTypes)
    }

    private val paramJavaTypes: Array<org.objectweb.asm.Type> =
            parameters
                .map { it.type }
                .map { it.toJavaType() }
                .toTypedArray()

    fun matches(call: MethodCall) =
            this.name == call.name && parameterTypesMatching(call)

    //TODO Also check call types, not only count
    private fun parameterTypesMatching(call: MethodCall) =
            this.parameters.size == call.parameters.size
}

data class Parameter(val name: String, val type: Type) {
    override fun toString() = "$name: $type"
}

